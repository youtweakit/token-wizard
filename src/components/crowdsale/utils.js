import { attachToContract } from '../../utils/blockchainHelpers'
import { noContractAlert } from '../../utils/alerts'
import { toFixed } from '../../utils/utils'
import { CONTRACT_TYPES } from '../../utils/constants'
import { contractStore, crowdsalePageStore, tokenStore, web3Store } from '../../stores'
import { toJS } from 'mobx'
import { BigNumber } from 'bignumber.js'

BigNumber.config({ DECIMAL_PLACES : 18 })

export const toBigNumber = (value) => isNaN(value) || value === '' ? new BigNumber(0) : new BigNumber(value)


export function getJoinedTiers(abi, addr, joinedCrowdsales, cb) {
  attachToContract(abi, addr)
    .then(crowdsaleContract => {
      console.log('attach to crowdsale contract')

      crowdsaleContract.methods.joinedCrowdsalesLen().call(function (err, joinedCrowdsalesLen) {
        if (err) {
          console.log(err)
          return cb([])
        }

        getJoinedTiersRecursively(0, crowdsaleContract, joinedCrowdsales, joinedCrowdsalesLen, function (_joinedCrowdsales) {
          cb(_joinedCrowdsales)
        })
      })
    })
    .catch(err => {
      console.log(err)
      cb([])
    })
}

function getJoinedTiersRecursively(i, crowdsaleContract, joinedCrowdsales, joinedCrowdsalesLen, cb) {
  if (i >= joinedCrowdsalesLen) {
    return cb(joinedCrowdsales);
  }

  crowdsaleContract.methods.joinedCrowdsales(i).call(function(err, joinedCrowdsale) {
    if (err) return console.log(err);
    console.log("joinedCrowdsale: " + joinedCrowdsale);

    if (joinedCrowdsale === "0x") {
      cb(joinedCrowdsales);
    } else {
      joinedCrowdsales.push(joinedCrowdsale);
      i++;
      getJoinedTiersRecursively(i, crowdsaleContract, joinedCrowdsales, joinedCrowdsalesLen, cb);
    }
  })
}

export function findCurrentContractRecursively(i, firstCrowdsaleContract, cb) {
  console.log(contractStore.crowdsale.addr);
  let crowdsaleAddr = contractStore.crowdsale.addr[i];
  const { web3 } = web3Store

  if (i === contractStore.crowdsale.addr.length) return cb(firstCrowdsaleContract, i);
  if (!crowdsaleAddr) return cb(null);
  if (!web3.utils.isAddress(crowdsaleAddr)) return cb(null);

  attachToContract(contractStore.crowdsale.abi, crowdsaleAddr)
    .then(crowdsaleContract => {
      console.log('attach to crowdsale contract')

      if (i === 0) {
        firstCrowdsaleContract = crowdsaleContract
      }

      if (!crowdsaleContract) return noContractAlert()

      console.log(contractStore.crowdsale.contractType)

      if (contractStore.crowdsale.contractType === CONTRACT_TYPES.standard) return cb(crowdsaleContract, i)

      crowdsaleContract.methods.startsAt().call(function (err, startDate) {
        if (err) return console.log(err)

        startDate = startDate * 1000
        console.log('startDate: ' + startDate)
        crowdsaleContract.methods.endsAt().call(function (err, endDate) {
          if (err) return console.log(err)

          endDate = endDate * 1000
          console.log('endDate: ' + endDate)

          let curDate = new Date().getTime()
          console.log('curDate: ' + curDate)
          if (curDate < endDate && curDate >= startDate) {
            cb(crowdsaleContract, i)
          } else {
            i++
            findCurrentContractRecursively(i, firstCrowdsaleContract, cb)
          }
        })
      })
    })
    .catch(err => console.log(err))
}

export function getCrowdsaleTargetDates($this, cb) {
  let propsCount = 0;
  let cbCount = 0;
  let state = $this.state;
  const { web3 } = web3Store

  for (let i = 0; i < contractStore.crowdsale.addr.length; i++) {
    let crowdsaleAddr = contractStore.crowdsale.addr[i];
    attachToContract(contractStore.crowdsale.abi, crowdsaleAddr)
      .then(crowdsaleContract => { // eslint-disable-line no-loop-func
        console.log("attach to crowdsale contract");

        if (!crowdsaleContract) return noContractAlert();

        if (crowdsaleContract.methods.startsAt) {
          propsCount++;
          crowdsaleContract.methods.startsAt().call((err, startDate) => {
            cbCount++;
            if (err) return console.log(err);

            console.log("startDate: " + startDate * 1000);
            if (!crowdsalePageStore.startDate || crowdsalePageStore.startDate > startDate * 1000)
              crowdsalePageStore.startDate = startDate * 1000;
            if (propsCount === cbCount) {
              state.loading = false;
              $this.setState(state, cb);
            }
          });
        }

        if (crowdsaleContract.methods.endsAt) {
          propsCount++;
          crowdsaleContract.methods.endsAt().call((err, endDate) => {
            cbCount++;
            if (err) return console.log(err);

            console.log("endDate: " + endDate * 1000);

            if (!crowdsalePageStore.endDate || crowdsalePageStore.endDate < endDate * 1000) {
              crowdsalePageStore.setProperty('endDate', endDate * 1000)
            }

            console.log("curDate: " + new Date().getTime());

            if (propsCount === cbCount) {
              state.loading = false;
              $this.setState(state, cb);
            }
          });
        }
      })
      .catch(console.log)
  }
}

export function initializeAccumulativeData() {
  crowdsalePageStore.setProperty('maximumSellableTokens', 0)
  crowdsalePageStore.setProperty('maximumSellableTokensInWei', 0)
  crowdsalePageStore.setProperty('investors', 0)
  crowdsalePageStore.setProperty('ethRaised', 0)
  crowdsalePageStore.setProperty('weiRaised', 0)
  crowdsalePageStore.setProperty('tokensSold', 0)
  crowdsalePageStore.setProperty('tokenAmountOf', 0)
  return Promise.resolve()
}

export function getAccumulativeCrowdsaleData() {
  const { web3 } = web3Store

  let promises = contractStore.crowdsale.addr
    .map(crowdsaleAddr => {
      return attachToContract(contractStore.crowdsale.abi, crowdsaleAddr)
        .then(crowdsaleContract => { // eslint-disable-line no-loop-func
          if (!crowdsaleContract) return noContractAlert()

          let getWeiRaised = crowdsaleContract.methods.weiRaised().call().then((weiRaised) => {
            let newWeiRaised
            if (crowdsalePageStore.weiRaised) {
              newWeiRaised = crowdsalePageStore.weiRaised + parseInt(weiRaised, 10);
            } else {
              newWeiRaised = parseInt(weiRaised, 10);
            }

            crowdsalePageStore.setProperty('weiRaised', newWeiRaised)
            crowdsalePageStore.setProperty('ethRaised', parseFloat(web3.utils.fromWei(toFixed(crowdsalePageStore.weiRaised).toString(), 'ether')))
          })

          let getTokensSold = crowdsaleContract.methods.tokensSold().call().then((tokensSold) => {
            if (crowdsalePageStore.tokensSold) {
              crowdsalePageStore.setProperty('tokensSold', crowdsalePageStore.tokensSold + parseInt(tokensSold, 10))
            } else {
              crowdsalePageStore.setProperty('tokensSold', parseInt(tokensSold, 10))
            }
          })

          let getMaximumSellableTokens = crowdsaleContract.methods.maximumSellableTokens().call().then((maximumSellableTokens) => {
            const maxSellableTokens = toBigNumber(crowdsalePageStore.maximumSellableTokens)
            crowdsalePageStore.setProperty('maximumSellableTokens', maxSellableTokens.plus(maximumSellableTokens).toFixed())

            //calc maximumSellableTokens in Eth
            return setMaximumSellableTokensInEth(crowdsaleContract, maximumSellableTokens)
          })

          let getInvestors = crowdsaleContract.methods.investorCount().call().then((investors) => {
            const oldInvestors = crowdsalePageStore.investors
            const investorsCount = parseInt(investors, 10)

            if (oldInvestors) {
              crowdsalePageStore.setProperty('investors', oldInvestors + investorsCount);
            } else {
              crowdsalePageStore.setProperty('investors', investorsCount);
            }
          });

          return Promise.all([getWeiRaised, getTokensSold, getMaximumSellableTokens, getInvestors])
        })
    })

  return Promise.all(promises)
}

function setMaximumSellableTokensInEth(crowdsaleContract, maximumSellableTokens) {
  return crowdsaleContract.methods.pricingStrategy().call()
    .then((pricingStrategyAddr) => {
      return attachToContract(contractStore.pricingStrategy.abi, pricingStrategyAddr)
    })
    .then(pricingStrategyContract => {
      if (!pricingStrategyContract) return noContractAlert()

      return pricingStrategyContract.methods.oneTokenInWei().call()
        .then((oneTokenInWei) => {
          const currentMaximumSellableTokensInWei = toBigNumber(crowdsalePageStore.maximumSellableTokensInWei)
          const maximumSellableTokensInWei = toBigNumber(oneTokenInWei).times(maximumSellableTokens).div(`1e${tokenStore.decimals}`).dp(0)

          crowdsalePageStore.setProperty('maximumSellableTokensInWei', currentMaximumSellableTokensInWei.plus(maximumSellableTokensInWei).toFixed())
        })
    })
}

export function getCurrentRate(crowdsaleContract) {
  return new Promise((resolve, reject) => {
    if (!crowdsaleContract) {
      noContractAlert()
      reject('no contract')
      return
    }

    crowdsaleContract.methods.pricingStrategy().call((err, pricingStrategyAddr) => {
      if (err) {
        console.log(err)
        reject(err)
        return
      }

      console.log('pricingStrategy:', pricingStrategyAddr)
      contractStore.setContractProperty('pricingStrategy', 'addr', pricingStrategyAddr)

      if (!pricingStrategyAddr || pricingStrategyAddr === "0x") {
        reject('no pricingStrategy address')
        return
      }

      getPricingStrategyData()
        .then(resolve)
        .catch(reject)
    });
  }
)}

export function getCrowdsaleData (crowdsaleContract) {
  return new Promise((resolve, reject) => {
    if (!crowdsaleContract) {
      noContractAlert()
      reject('no contract')
      return
    }

    console.log(crowdsaleContract)

    let propsCount = 0
    let cbCount = 0

    propsCount++
    crowdsaleContract.methods.token().call((err, tokenAddr) => {
      cbCount++

      if (err) {
        return console.log(err)
      }

      console.log('token:', tokenAddr)
      contractStore.setContractProperty('token', 'addr', tokenAddr)

      // if (propsCount === cbCount) {
      //   resolve()
      //   return
      // }

      if (!tokenAddr || tokenAddr === '0x') {
        return console.log('no token address available:', tokenAddr)
      }

      propsCount++
      getTokenData()
        .then(() => {
          if (!crowdsaleContract.methods.pricingStrategy) {
            reject('no pricingStrategy method')
            return
          }

          propsCount++
          crowdsaleContract.methods.pricingStrategy().call((err, pricingStrategyAddr) => {
            cbCount++

            if (err) {
              reject(err)
              return console.log(err)
            }

            console.log('pricingStrategy:', pricingStrategyAddr)
            contractStore.setContractProperty('pricingStrategy', 'addr', pricingStrategyAddr)

            if (propsCount === cbCount) {
              resolve()
              return
            }

            if (!pricingStrategyAddr || pricingStrategyAddr === '0x') {
              reject('no pricingStrategyAddr')
              return
            }

            getPricingStrategyData()
              .then(() => {
                cbCount++
                if (propsCount === cbCount) {
                  resolve()
                }
              })
              .catch(reject)
          })
        })
        .catch(reject)
    })
  })
}

function getTokenData () {
  return new Promise((resolve, reject) => {
    const { web3 } = web3Store

    if (!web3) {
      resolve('no MetaMask')
      return
    }

    web3.eth.getAccounts().then(accounts => {
      if (accounts.length === 0) {
        resolve('no accounts')
        return
      }

      let propsCount = 0
      let cbCount = 0
      let tokenObj = toJS(contractStore.token)
      console.log('tokenObj', tokenObj)

      attachToContract(tokenObj.abi, tokenObj.addr)
        .then(tokenContract => {
          console.log('attach to token contract')

          if (!tokenContract) {
            noContractAlert()
            reject('no contract')
            return
          }

          propsCount++
          tokenContract.methods.name().call((err, name) => {
            cbCount++

            if (err) {
              return console.log(err)
            }

            console.log('token name:', name)
            tokenStore.setProperty('name', name)

            if (propsCount === cbCount) {
              resolve()
            }
          })

          propsCount++
          tokenContract.methods.symbol().call((err, ticker) => {
            cbCount++

            if (err) {
              console.log(err)
            }

            console.log('token ticker: ' + ticker)
            tokenStore.setProperty('ticker', ticker)

            if (propsCount === cbCount) {
              resolve()
            }
          })

          if (tokenContract.methods.balanceOf) {
            propsCount++
            tokenContract.methods.balanceOf.call(accounts[0], (err, balanceOf) => {
              cbCount++

              if (err) {
                return console.log(err)
              }

              console.log('balanceOf: ' + balanceOf)
              const tokenAmountOf = crowdsalePageStore.tokenAmountOf ? crowdsalePageStore.tokenAmountOf : 0
              crowdsalePageStore.setProperty('tokenAmountOf', tokenAmountOf + parseInt(balanceOf, 10))

              if (propsCount === cbCount) {
                resolve()
              }
            })
          }

          propsCount++
          tokenContract.methods.decimals().call((err, decimals) => {
            cbCount++

            if (err) {
              console.log(err)
            }

            console.log('token decimals:', decimals)
            tokenStore.setProperty('decimals', decimals)

            if (propsCount === cbCount) {
              resolve()
            }
          })

          propsCount++
          tokenContract.methods.totalSupply().call((err, supply) => {
            cbCount++

            if (err) {
              console.log(err)
            }

            console.log('token supply:', supply)
            tokenStore.setProperty('supply', supply)

            if (propsCount === cbCount) {
              resolve()

            } else {
              let propsCount = 0
              let cbCount = 0

              attachToContract(contractStore.token.abi, contractStore.token.addr)
                .then(tokenContract => {
                  console.log('attach to token contract')

                  if (!tokenContract) {
                    noContractAlert()
                    reject('no contract')
                    return
                  }

                  propsCount++
                  tokenContract.methods.name().call((err, name) => {
                    cbCount++

                    if (err) {
                      return console.log(err)
                    }

                    console.log('token name:', name)
                    tokenStore.setProperty('name', name)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })

                  propsCount++
                  tokenContract.methods.symbol().call((err, ticker) => {
                    cbCount++

                    if (err) {
                      console.log(err)
                    }

                    console.log('token ticker:', ticker)
                    tokenStore.setProperty('ticker', ticker)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })

                  if (tokenContract.methods.balanceOf) {
                    propsCount++
                    tokenContract.methods.balanceOf(accounts[0]).call((err, balanceOf) => {
                      cbCount++

                      if (err) {
                        return console.log(err)
                      }

                      console.log('balanceOf:', balanceOf)
                      const tokenAmountOf = crowdsalePageStore.tokenAmountOf ? crowdsalePageStore.tokenAmountOf : 0
                      crowdsalePageStore.setProperty('tokenAmountOf', tokenAmountOf + parseInt(balanceOf, 10))

                      if (propsCount === cbCount) {
                        resolve()
                      }
                    })
                  }

                  propsCount++
                  tokenContract.methods.decimals().call((err, decimals) => {
                    cbCount++

                    if (err) {
                      console.log(err)
                    }

                    console.log('token decimals:', decimals)
                    tokenStore.setProperty('decimals', decimals)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })

                  propsCount++
                  tokenContract.methods.totalSupply().call((err, supply) => {
                    cbCount++

                    if (err) {
                      console.log(err)
                    }

                    console.log('token supply:', supply)
                    tokenStore.setProperty('supply', supply)

                    if (propsCount === cbCount) {
                      resolve()
                    }
                  })
                })
                .catch(reject)
            }
          })
        })
        .catch(reject)
    })
  })
}

export function getPricingStrategyData () {
  return new Promise((resolve, reject) => {
    attachToContract(contractStore.pricingStrategy.abi, contractStore.pricingStrategy.addr)
      .then(pricingStrategyContract => {
        console.log('attach to pricing strategy contract')

        if (!pricingStrategyContract) {
          noContractAlert()
          reject('no contract')
          return
        }

        pricingStrategyContract.methods.oneTokenInWei().call((err, rate) => {
          if (err) {
            console.log(err)
          }

          console.log('pricing strategy rate:', rate)
          crowdsalePageStore.setProperty('rate', rate)
          resolve()
        })
      })
      .catch(err => {
        reject(err)
        console.log(err)
      })
  })
}

export const getContractStoreProperty = (contract, property) => {
  const text = contractStore && contractStore[contract] && contractStore[contract][property]
  return text === undefined ? '' : text
}
