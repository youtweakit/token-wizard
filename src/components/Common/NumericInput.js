import React, { Component } from 'react'
import { InputField } from './InputField'
import { VALIDATION_TYPES } from '../../utils/constants'
import { countDecimalPlaces } from '../../utils/utils'
import { observer } from 'mobx-react'

const { VALID, INVALID } = VALIDATION_TYPES

@observer
export class NumericInput extends Component {
  constructor (props) {
    super(props)

    this.state = {
      value: props.value || '',
      pristine: props.pristine !== undefined ? props.pristine : true,
      valid: props.valid || VALID
    }
  }

  onPaste = e => {
    if (isNaN(parseFloat(e.clipboardData.getData('text/plain')))) {
      e.preventDefault()
    }
  }

  onKeyPress = e => {
    const { min, max, acceptFloat } = this.props
    const allowsNegative = min < 0 || max < 0

    if ((e.key === 'e' || e.key === '.' || e.key === '+') && !acceptFloat) {
      e.preventDefault()
    }

    // '-' symbol is required for scientific notation and negative values
    if (e.key === '-' && !allowsNegative && !acceptFloat) {
      e.preventDefault()
    }
  }

  onChange = e => {
    const value = this.props.acceptFloat ? parseFloat(e.target.value) : parseInt(e.target.value, 10)
    this.validate(value)
  }

  validate = value => {
    const { acceptFloat, maxDecimals, minDecimals, min, max } = this.props
    let isValid = true

    if (acceptFloat) {
      if (maxDecimals !== undefined) {
        isValid = isValid && countDecimalPlaces(value) <= maxDecimals
      }

      if (minDecimals !== undefined) {
        isValid = isValid && countDecimalPlaces(value) >= minDecimals
      }
    }

    if (isValid && min !== undefined) {
      isValid = isValid && value >= min
    }

    if (isValid && max !== undefined) {
      isValid = isValid && value <= max
    }

    const result = {
      value: '',
      pristine: value === this.props.value && this.state.pristine,
      valid: INVALID
    }

    if (!isNaN(value)) {
      result.value = value
      result.valid = isValid ? VALID : INVALID
    }

    this.props.onValueUpdate(result)
  }

  componentWillReceiveProps (newProps) {
    this.setState({
      value: newProps.value,
      pristine: newProps.pristine,
      valid: newProps.valid
    })
  }

  componentDidUpdate (prevProps) {
    const { acceptFloat, maxDecimals, minDecimals, min, max } = this.props

    if (
      prevProps.acceptFloat !== acceptFloat ||
      prevProps.minDecimals !== minDecimals ||
      prevProps.maxDecimals !== maxDecimals ||
      prevProps.min !== min ||
      prevProps.max !== max
    ) {
      // re-check validity if any of the props had changed
      this.validate(this.state.value)
    }
  }

  render () {
    const { value, pristine, valid } = this.state
    const { disabled, side, errorMessage, title, description } = this.props

    return (
      <InputField
        disabled={disabled}
        side={side}
        type='number'
        errorMessage={errorMessage}
        value={value}
        pristine={pristine}
        valid={valid}
        title={title}
        onKeyPress={this.onKeyPress}
        onChange={this.onChange}
        onPaste={this.onPaste}
        description={description}
      />
    )
  }
}
