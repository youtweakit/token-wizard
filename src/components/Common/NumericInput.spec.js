import React from 'react'
import { NumericInput } from './NumericInput'
import { TEXT_FIELDS, VALIDATION_MESSAGES, VALIDATION_TYPES } from '../../utils/constants'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-15'
import { configure, mount } from 'enzyme'

configure({ adapter: new Adapter() })

describe('NumericInput', () => {
  const INPUT_EVENT = {
    KEYPRESS: 'keypress',
    CHANGE: 'change',
    PASTE: 'paste'
  }

  let changeMock
  let keypressMock
  let pasteMock
  let numericInputComponent
  let wrapperMemo, wrapper
  let inputMemo, input

  beforeEach(() => {
    changeMock = { target: { value: '' } }

    keypressMock = { key: '1', preventDefault: jest.fn() }

    pasteMock = {
      preventDefault: jest.fn(),
      clipboardData: {
        getData: () => 'e123e123'
      }
    }

    numericInputComponent = {
      side: 'left',
      title: TEXT_FIELDS.DECIMALS,
      description:
        'Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous).',
      errorMessage: VALIDATION_MESSAGES.DECIMALS,
      onValueUpdate: jest.fn()
    }

    wrapperMemo = undefined
    wrapper = () => wrapperMemo || (wrapperMemo = mount(React.createElement(NumericInput, numericInputComponent)))

    inputMemo = undefined
    input = () => inputMemo || (inputMemo = wrapper().find('input').at(0))
  })

  it('Should render the component', () => {
    numericInputComponent.min = 0
    numericInputComponent.max = 18
    numericInputComponent.acceptFloat = true
    numericInputComponent.minDecimals = 0
    numericInputComponent.maxDecimals = 4

    expect(renderer.create(React.createElement(NumericInput, numericInputComponent)).toJSON()).toMatchSnapshot()
  })
  it('Should prevent pasting invalid value', () => {
    input().simulate(INPUT_EVENT.PASTE, pasteMock)
    expect(pasteMock.preventDefault).toHaveBeenCalled()
  })
  it('Should allow pasting valid value', () => {
    pasteMock.clipboardData.getData = () => '12'

    input().simulate(INPUT_EVENT.PASTE, pasteMock)
    expect(pasteMock.preventDefault).toHaveBeenCalledTimes(0)
  })
  it('Should call onValueUpdate callback on successful update', () => {
    changeMock.target.value = '10'

    input().simulate(INPUT_EVENT.CHANGE, changeMock)
    expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
    expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
      value: 10,
      pristine: false,
      valid: VALIDATION_TYPES.VALID
    })
  })

  describe('symbols', () => {
    it('Should consider empty string as valid', () => {
      input().simulate(INPUT_EVENT.CHANGE, changeMock)
      expect(numericInputComponent.onValueUpdate).toHaveBeenCalled()
    })

    describe('min', () => {
      it('Should accept "-" symbol if min is negative', () => {
        numericInputComponent.min = -10
        keypressMock.key = '-'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(0)
      })
      it('Should accept "-" symbol if min is negative and accepts floating numbers', () => {
        numericInputComponent.min = -10
        numericInputComponent.acceptFloat = true
        keypressMock.key = '-'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(0)
      })
      it('Should reject "-" symbol if min is not negative', () => {
        numericInputComponent.min = 0
        keypressMock.key = '-'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalled()
      })
    })

    describe('max', () => {
      it('Should accept "-" symbol if max is negative', () => {
        numericInputComponent.max = -10
        keypressMock.key = '-'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(0)
      })
      it('Should accept "-" symbol if max is negative and accepts floating numbers', () => {
        numericInputComponent.max = -10
        numericInputComponent.acceptFloat = true
        keypressMock.key = '-'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(0)
      })
      it('Should reject "-" symbol if max is not negative', () => {
        numericInputComponent.max = 10
        keypressMock.key = '-'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalled()
      })
    })

    it('Should prevent "." if no float is allowed', () => {
      numericInputComponent.value = '10'
      keypressMock.key = '.'

      input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
      expect(keypressMock.preventDefault).toHaveBeenCalledTimes(1)
    })

    it('Should prevent "+" if no float is allowed', () => {
      numericInputComponent.value = '10'
      keypressMock.key = '+'

      input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
      expect(keypressMock.preventDefault).toHaveBeenCalledTimes(1)
    })

    it('Should prevent "e" if no float is allowed', () => {
      numericInputComponent.value = '10'
      keypressMock.key = 'e'

      input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
      expect(keypressMock.preventDefault).toHaveBeenCalledTimes(1)
    })
  })

  describe('integer numbers', () => {
    describe('min', () => {
      it('Should fail if value is lesser than min', () => {
        numericInputComponent.min = 5
        changeMock.target.value = '4'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 4,
          pristine: false,
          valid: VALIDATION_TYPES.INVALID
        })
      })
      it('Should pass if value is greater than min', () => {
        numericInputComponent.min = 5
        changeMock.target.value = '8'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 8,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
    })

    describe('max', () => {
      it('Should fail if value is greater than max', () => {
        numericInputComponent.max = 15
        changeMock.target.value = '20'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 20,
          pristine: false,
          valid: VALIDATION_TYPES.INVALID
        })
      })
      it('Should pass if value is lesser than min', () => {
        numericInputComponent.max = 15
        changeMock.target.value = '10'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 10,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
    })
  })

  describe('float numbers', () => {
    beforeEach(() => {
      numericInputComponent.acceptFloat = true
      numericInputComponent.minDecimals = 2
      numericInputComponent.maxDecimals = 4
    })

    describe('maxDecimals', () => {
      it('Should fail if value has more decimals', () => {
        changeMock.target.value = '1.12345'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.12345,
          pristine: false,
          valid: VALIDATION_TYPES.INVALID
        })
      })
      it('Should pass if value has less decimals', () => {
        changeMock.target.value = '1.123'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.123,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
      it('Should pass if value has same decimals', () => {
        changeMock.target.value = '1.1234'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.1234,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
    })

    describe('minDecimals', () => {
      it('Should fail if value has less decimals', () => {
        changeMock.target.value = '1.1'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.1,
          pristine: false,
          valid: VALIDATION_TYPES.INVALID
        })
      })
      it('Should pass if value has more decimals', () => {
        changeMock.target.value = '1.1234'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.1234,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
      it('Should pass if value has same decimals', () => {
        changeMock.target.value = '1.12'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.12,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
    })

    describe('range [minDecimals, maxDecimals]', () => {
      it('Should fail for: 1.1, if it is outside the range', () => {
        changeMock.target.value = '1.1'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.1,
          pristine: false,
          valid: VALIDATION_TYPES.INVALID
        })
      })
      it('Should fail for: 1.12345, if it is outside the range', () => {
        changeMock.target.value = '1.12345'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.12345,
          pristine: false,
          valid: VALIDATION_TYPES.INVALID
        })
      })
      it('Should pass for: 1.12, if it is inside the range', () => {
        changeMock.target.value = '1.12'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.12,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
      it('Should pass for: 1.123, if it is inside the range', () => {
        changeMock.target.value = '1.123'

        input().simulate(INPUT_EVENT.CHANGE, changeMock)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledTimes(1)
        expect(numericInputComponent.onValueUpdate).toHaveBeenCalledWith({
          value: 1.123,
          pristine: false,
          valid: VALIDATION_TYPES.VALID
        })
      })
    })

    describe('dot notation', () => {
      it('Should reject float with dot notation', () => {
        numericInputComponent.acceptFloat = false
        numericInputComponent.value = 1

        keypressMock.key = '.'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalled()
      })
      it('Should accept float with dot notation', () => {
        numericInputComponent.minDecimals = 0
        numericInputComponent.value = 1

        keypressMock.key = '.'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(0)
        expect()
      })
    })

    describe('scientific notation', () => {
      it('Should reject float with scientific notation', () => {
        numericInputComponent.acceptFloat = false
        numericInputComponent.value = 1

        keypressMock.key = 'e'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalled()
      })
      it('Should accept float with scientific notation', () => {
        numericInputComponent.value = 1

        keypressMock.key = 'e'

        input().simulate(INPUT_EVENT.KEYPRESS, keypressMock)
        expect(keypressMock.preventDefault).toHaveBeenCalledTimes(0)
      })
    })
  })
})
