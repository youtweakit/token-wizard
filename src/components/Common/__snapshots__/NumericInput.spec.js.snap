// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput Should render the component 1`] = `
<div
  className="left"
>
  <label
    className="label"
  >
    Decimals
  </label>
  <input
    className="input"
    disabled={undefined}
    onBlur={undefined}
    onChange={[Function]}
    onKeyPress={[Function]}
    onPaste={[Function]}
    style={undefined}
    type="number"
    value=""
  />
  <p
    className="description"
  >
    Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous).
  </p>
  <p
    style={
      Object {
        "color": "red",
        "fontSize": "12px",
        "fontWeight": "bold",
        "height": "10px",
        "width": "100%",
      }
    }
  />
</div>
`;
